src/HCubature.jl

```diff
@@ -46,8 +46,24 @@ end
 cubrule(::Val{0}, ::Type{T}) where {T} = Trivial()
 countevals(::Trivial) = 1
 
-function hcubature_(f::F, a::SVector{n,T}, b::SVector{n,T}, norm, rtol_, atol,
-                    maxevals, initdiv) where {F, n, T<:Real}
+"""
+    alloc_buf(;dimension[, domain_type, range_type, error_type])
+
+Allocate a buffer that can be used in calls to [`hcubature`](@ref).
+
+# Examples:
+
+```julia
+ buffer = alloc_buf(;dimension=2, range_type=ComplexF64, domain_type=Float64)
+ I, E = hcubature(x -> 2+im, (0,0), (2pi, pi); buffer))
+ ```
+
+"""
+function alloc_buf(;dimension, domain_type=Float64, range_type=Float64, error_type=real(range_type))
```

Comment:
It would be good to provide a type-stable interface, e.g. where `dimension` is passed as a `::Val{dimension}`, and the types should be passed as `::Type{T}` arguments.

Maybe call it `hcubature_buffer`?

Maybe accept same arguments as `hcubature` to make life simple (and then just evaluate the function once to get the range type … this API is only useful in cases where the integrand is relatively cheap, after all).

Reply:
Thanks for the suggestions. I renamed the function to `hcubature_buffer` and implemented the suggested API to mimic `hcubature` (it should now be type-stable):

```julia
f = x->cos(x[1])
a,b = (0,), (1,)
buffer = hcubature_buffer(f,a,b)
I,E = hcubature(f,a,b;buffer)
```

The buffer constructor taking a `Val{dimension}` and `::Type{T}` arguments seemed a bit cumbersome to me. Since avoiding one evaluation is unimportant if the goal is to reuse the buffer several times, I decided not to implement that. Is this OK?

Reply:
Seems good to me.